### PR TITLE
Forever promote02

### PR DESCRIPTION
**Pull request by miketang84**
promote some of the forever behaves.

### CODE REVIEW

**Comment by TheWaWaR** [comment]
@daogangtang 
```
git config --global user.email "email@example.com"
```

https://help.github.com/articles/setting-your-commit-email-address-in-git/#setting-your-email-address-for-every-repository-on-your-computer

File: src/main.rs [hunk 1]
```diff
@@ -32,8 +32,9 @@ use config::ForeverConfig;
 use process::Processes;
 
 use std::env;
-use std::thread;
-use std::time;
+//use std::process::exit;
+//use std::thread;
+//use std::time;
```
Comment (TheWaWaR) [inline src/main.rs]: Just delete those lines

File: src/process.rs [hunk 2]
```diff
@@ -270,9 +291,21 @@ fn check_process(pidfile: String) -> Option<u32> {
...
-            Err(_) => None,
+        let pid_str = pid.to_string();
+        let args = vec!["-p", &pid_str];
+        let output = Command::new("ps")
+            .args(args)
+            .output()
+            .expect("failed to execute ps -p");
+        if output.status.success() {
+            let otpt_str = String::from_utf8(output.stdout).unwrap();
+            if otpt_str.contains(&pid.to_string()) {
```
Comment (TheWaWaR) [inline src/process.rs:300]: `contains` is incorrect?

For example:
* `pid` = `163`
* `otpt_str` = `121631`

File: src/main.rs [hunk 3]
```diff
@@ -102,9 +103,6 @@ fn main() {
         Some(&_) => {}
         None => {
             daemon.start_all();
-            loop {
-                thread::sleep(time::Duration::from_secs(1))
-            }
```
Comment (keroro520) [inline src/main.rs:107]: 这段去掉和不去掉的区别是？

File: src/process.rs [hunk 4]
```diff
@@ -95,8 +101,16 @@ impl Processes {
 
     // run all child processes
     pub fn start_all(self) {
-        for (_, child_process) in self.children {
-            run_process(child_process);
+        let (tx, rx): (Sender<String>, Receiver<String>) = channel();
+
+        for (_, child_process) in self.children.clone() {
```
Comment (keroro520) [inline src/process.rs]: 可能不需要 `.clone()` 

或者 `for child_process in self.children.values() ` 更好

File: src/process.rs [hunk 5]
```diff
@@ -203,7 +218,13 @@ pub fn run_process(child_process: Arc<Mutex<Processes>>) {
                     }
                 }
             }
+
             if !change_status(&child_process) {
+                // inform cita-forever error occurs.
+                let name = child_process.lock().processcfg.name.clone().unwrap();
+                tx.send(format!("==>: Child process {} exited unexpectly!", name))
```
Comment (keroro520) [inline src/process.rs]: unexpectly => unexpectedly

File: src/process.rs [hunk 6]
```diff
@@ -203,7 +218,13 @@ pub fn run_process(child_process: Arc<Mutex<Processes>>) {
...
+                    .unwrap();
+                thread::sleep(time::Duration::from_secs(1));
```
Comment (keroro520) [inline src/process.rs]: why sleep here ?

File: src/process.rs (hunk 2 above)
Comment (miketang84) [reply src/process.rs:300]: That's correct, because `ps -p` returns only the info of that process, no other process.

Comment (TheWaWaR) [reply src/process.rs:300]: ok, maybe `ps -p {pid} -o pid=` is more clean way.

File: src/main.rs (hunk 1 above)
Comment (miketang84) [reply src/main.rs]: ok

**Comment by miketang84** [comment]
just do another pr to merge those trivial comments.

**Review by keroro520 — COMMENTED** [review]
> just do another pr to merge those trivial comments.

**Comment by keroro520** [comment]
> just do another pr to merge those trivial comments.

I think you should rebase these commits and then re-push, but not open another PR.

File: src/main.rs (hunk 3 above)
Comment (miketang84) [reply src/main.rs:107]: start_all中已经做了一个loop，外面这个就没必要了。

File: src/process.rs (hunk 2 above)
Comment (miketang84) [reply src/process.rs:300]: 已经改好，并测试完毕。

**Review by u2 — APPROVED** [review]

**Pull request merged**